File: metal/paddle-mobile/paddle-mobile/Src/Operators/MulticlassNMSOp.swift
/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License. */

import Foundation

class MulticlassNMSParam<P: PrecisionProtocol>: OpParam {
    //typealias ParamPrecisionType = P
    required init(opDesc: PMOpDesc, inScope: Scope) throws {
        scores = try MulticlassNMSParam.getFirstTensor(key: "Scores", map: opDesc.inputs, from: inScope)
        bboxes = try MulticlassNMSParam.getFirstTensor(key: "BBoxes", map: opDesc.inputs, from: inScope)
        output = try MulticlassNMSParam.outputOut(outputs: opDesc.outputs, from: inScope)
        
        middleOutput = FetchHolder.init(inPaddedCapacity: scores.tensorDim.numel(), inDim: scores.tensorDim)
        
        bboxOutput = FetchHolder.init(inPaddedCapacity: bboxes.tensorDim.numel(), inDim: bboxes.tensorDim)
    }
    var bboxOutput: FetchHolder
    var middleOutput: FetchHolder
    let scores: Texture
    let bboxes: Texture
    var output: Texture
}

class MulticlassNMSOp<P: PrecisionProtocol>: Operator<MulticlassNMSKernel<P>, MulticlassNMSParam<P>>, Runable, Creator, InferShaperable{
    
    func inputVariant() -> [String : [MTLBuffer]]? {
        guard let scoreBuffer = para.middleOutput.resultBuffer, let bboxBuffer = para.middleOutput.resultBuffer else {
            return nil
        }
        return ["Scores" : [scoreBuffer], "BBoxes" : [bboxBuffer]]
    }
    
    func computeMiddleResult(device: MTLDevice, buffer: MTLCommandBuffer) throws {
        try kernel.compute(commandBuffer: buffer, param: para)
    }
    
    func inferShape() {
        // para.output.dim = para.input.dim
    }
    
    typealias OpType =  MulticlassNMSOp<P>
    func runImpl(device: MTLDevice, buffer: MTLCommandBuffer) throws {
        
    }
    
    func delogOutput() {
        print(" nms - output: ")
        do {
            let output = try para.bboxes.metalTexture?.float32Array().strideArray() ?? []
            print(output)
        } catch _ {
        }
    }
}

File: mobile/src/operators/multiclass_nms_op.cpp
/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

#ifdef MULTICLASSNMS_OP

#include "operators/multiclass_nms_op.h"
namespace paddle_mobile {
namespace operators {

template <typename Dtype, typename T>
void MultiClassNMSOp<Dtype, T>::InferShape() const {
  auto input_bboxes_dims = this->param_.InputBBoxes()->dims();
  auto input_scores_dims = this->param_.InputScores()->dims();
  if (input_scores_dims.size() != 3) {
    LOG(kLOG_ERROR) << "Input Scores size must be 3";
  }
  if (input_bboxes_dims[2] % 4 != 0 || input_bboxes_dims[2] < 4) {
    LOG(kLOG_ERROR) << "Input BBoxes 2nd dimension must be multiples of 4";
  }
  if (input_bboxes_dims[1] != input_scores_dims[2]) {
    LOG(kLOG_ERROR) << "Predict bboxes must be equal";
  }
  // pre size, will change in Compute.
  this->param_.Out()->Resize(
      framework::make_ddim({input_bboxes_dims[1], input_bboxes_dims[2] + 2}));
}

}  // namespace operators
}  // namespace paddle_mobile

namespace ops = paddle_mobile::operators;
#ifdef PADDLE_MOBILE_CPU
REGISTER_OPERATOR_CPU(multiclass_nms, ops::MultiClassNMSOp);
#endif
#ifdef PADDLE_MOBILE_CL
REGISTER_OPERATOR_CL(multiclass_nms, ops::MultiClassNMSOp);
#endif

#endif
